File: bbgm/util/lock.py
"""
bbgm.util.lock

These functions all deal with locking game state when there is some blocking
action in progress, currently one of these things:

* Game simulation is in progress
* User is negotiating a contract

There are also functions to check if it is permissible to start one of those
actions.
"""

from flask import g

def set_games_in_progress(status):
    g.dbex('UPDATE game_attributes SET games_in_progress = :games_in_progress', games_in_progress=status)

def games_in_progress():
    r = g.dbex('SELECT games_in_progress FROM game_attributes')
    in_progress, = r.fetchone()
    return in_progress

def negotiation_in_progress():
    """Returns True or False depending on whether the negotiations table is
    empty or not."""
    r = g.dbex('SELECT 1 FROM negotiations')
    if r.rowcount:
        return True
    else:
        return False

def can_start_games():
    """Returns a boolean. Games can be started only when there is no contract
    negotiation in progress and there is no other game simulation in progress.
    """
    r = g.dbex('SELECT games_in_progress FROM game_attributes')
    games_in_progress, = r.fetchone()

    if games_in_progress or negotiation_in_progress():
        return False
    else:
        return True

def can_start_negotiation():
    r = g.dbex('SELECT games_in_progress FROM game_attributes')
    games_in_progress, = r.fetchone()

    if games_in_progress:
        return False

    # Allow multiple parallel negotiations (ignore negotiation_in_progress) only for resigning players
    r = g.dbex('SELECT 1 FROM negotiations WHERE resigning = FALSE')
    if r.rowcount:
        return False

    return True

File: src/js/util/lock.js
// @flow

import g from '../globals';
import * as league from '../core/league';
import {getCopy} from '../db';

/**
 * Is game simulation in progress?
 *
 * Calls the callback function with either true or false depending on whether there is a game simulation currently in progress.
 *
 * @memberOf util.lock
 * @return {Promise.boolean}
 */
async function gamesInProgress(): Promise<boolean> {
    await league.loadGameAttribute("gamesInProgress");
    return g.gamesInProgress;
}

/**
 * Is a negotiation in progress?
 *
 * Calls the callback function with either true or false depending on whether there is an ongoing negoation.
 *
 * @memberOf util.lock
 * @return {Promise.boolean}
 */
async function negotiationInProgress(): Promise<boolean> {
    const negotiations = await g.cache.getAll('negotiations');
    return negotiations.length > 0;
}

/**
 * Is a phase change in progress?
 *
 * @memberOf util.lock
 * @return {Promise.boolean}
 */
async function phaseChangeInProgress(): Promise<boolean> {
    await league.loadGameAttribute("phaseChangeInProgress");
    return g.phaseChangeInProgress;
}

/**
 * Can new game simulations be started?
 *
 * Calls the callback function with either true or false. If games are in progress or any contract negotiation is in progress, false.
 *
 * @memberOf util.lock
 * @return {Promise.boolean}
 */
async function canStartGames(): Promise<boolean> {
    const gamesInProgressBool = await gamesInProgress();
    if (gamesInProgressBool) {
        return false;
    }

    const negotiationInProgressBool = await negotiationInProgress();
    if (negotiationInProgressBool) {
        return false;
    }

    const phaseChangeInProgressBool = await phaseChangeInProgress();
    if (phaseChangeInProgressBool) {
        return false;
    }

    return true;
}

/**
 * Can a new contract negotiation be started?
 *
 * Calls the callback function with either true or false. If games are in progress or a free agent (not re-signing!) is being negotiated with, false.
 *
 * @memberOf util.lock
 * @return {Promise.boolean}
 */
async function canStartNegotiation(): Promise<boolean> {
    const gamesInProgressBool = await gamesInProgress();
    if (gamesInProgressBool) {
        return false;
    }

    // Allow multiple parallel negotiations only for re-signing players
    const negotiations = await g.cache.getAll('negotiations');
    for (const negotiation of negotiations) {
        if (!negotiation.resigning) {
            return false;
        }
    }

    return true;

    // Don't also check phase change because negotiations are auto-started in phase change
}

/**
 * Is there an undread message from the owner?
 *
 * Calls the callback function with either true or false.
 *
 * @memberOf util.lock
 * @return {Promise.boolean}
 */
async function unreadMessage(): Promise<boolean> {
    const messages = await getCopy.messages();
    for (let i = 0; i < messages.length; i++) {
        if (!messages[i].read) {
            return true;
        }
    }
    return false;
}

export {
    gamesInProgress,
    negotiationInProgress,
    phaseChangeInProgress,
    canStartGames,
    canStartNegotiation,
    unreadMessage,
};

